x/staking/keeper/cons_pubkey.go

```diff
@@ -0,0 +1,95 @@
+package keeper
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/x/staking/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// maxRotations is the value of max rotations can be made in unbonding period for a validator.
+const maxRotations = 1
+
+// setConsPubKeyRotationHistory sets the consensus key rotation of a validator into state
+func (k Keeper) setConsPubKeyRotationHistory(
+	ctx context.Context, valAddr sdk.ValAddress,
+	oldPubKey, newPubKey *codectypes.Any, fee sdk.Coin,
+) error {
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	height := uint64(sdkCtx.BlockHeight())
+	history := types.ConsPubKeyRotationHistory{
+		OperatorAddress: valAddr.Bytes(),
+		OldConsPubkey:   oldPubKey,
+		NewConsPubkey:   newPubKey,
+		Height:          height,
+		Fee:             fee,
+	}
+	err := k.ValidatorConsPubKeyRotationHistory.Set(ctx, collections.Join(valAddr.Bytes(), height), history)
+	if err != nil {
+		return err
+	}
+
+	if err := k.BlockConsPubKeyRotationHistory.Set(ctx, height, history); err != nil {
+		return err
+	}
+
+	ubdTime, err := k.UnbondingTime(ctx)
+	if err != nil {
+		return err
+	}
+
+	queueTime := sdkCtx.BlockHeader().Time.Add(ubdTime)
+	if err := k.ValidatorConsensusKeyRotationRecordIndexKey.Set(ctx, collections.Join(valAddr.Bytes(), queueTime), []byte{}); err != nil {
+		return err
+	}
+
+	return k.SetConsKeyQueue(ctx, queueTime, valAddr)
+}
+
+// exceedsMaxRotations returns true if the key rotations exceed the limit, currently we are limiting one rotation for unbonding period.
+func (k Keeper) exceedsMaxRotations(ctx context.Context, valAddr sdk.ValAddress) (bool, error) {
+	count := 0
+	rng := collections.NewPrefixUntilPairRange[[]byte, time.Time](valAddr)
```

Comment:
This will also prefix over other val addresses, `PrefixUntil`  means that we iterate until the provided prefix starting from the first key in the collection.

If you want to prefix only over val addr then you should be using: PrefixedPairRange.

If this is wrong, and it was uncaught by tests then we should add a test which assesses this behaviour.